# Using Python for Research Homework: Week 1

In this homework, we will use objects, functions, and randomness to find the length of documents, approximate $\pi$, and smooth out random noise.

### Exercise 1

In this five-part exercise, we will count the frequency of each letter in a given string.

#### Exercise 1a

- Import the `string` library.
- Create a variable `alphabet` that consists of the lowercase and uppercase letters in the English alphabet using the `ascii_letters` data attribute of the `string` library.

In [1]:
import string
alphabet = string.ascii_letters

#### Exercise 1b
- The lower and upper case letters of the English alphabet should stored as the string variable `alphabet`.
- Consider the sentence 'Jim quickly realized that the beautiful gowns are expensive'. Create a dictionary `count_letters` with keys consisting of each unique letter in the sentence and values consisting of the number of times each letter is used in this sentence. Count upper case and lower case letters separately in the dictionary.

In [2]:
count_letters = dict()

for letter in alphabet:
    count = 'Jim quickly realized that the beautiful gowns are expensive'.count(letter)
    if count > 0:
        count_letters[letter] = count
    else:
        pass

In [3]:
count_letters

{'a': 4,
 'b': 1,
 'c': 1,
 'd': 1,
 'e': 8,
 'f': 1,
 'g': 1,
 'h': 2,
 'i': 5,
 'k': 1,
 'l': 3,
 'm': 1,
 'n': 2,
 'o': 1,
 'p': 1,
 'q': 1,
 'r': 2,
 's': 2,
 't': 4,
 'u': 3,
 'v': 1,
 'w': 1,
 'x': 1,
 'y': 1,
 'z': 1,
 'J': 1}

In [4]:
count_letters.values()

dict_values([4, 1, 1, 1, 8, 1, 1, 2, 5, 1, 3, 1, 2, 1, 1, 1, 2, 2, 4, 3, 1, 1, 1, 1, 1, 1])

In [5]:
sentence = "Jim quickly realized that the beautiful gowns are expensive"

Unique_letters = set("Jim quickly realized that the beautiful gowns are expensive")
Unique_letters = list(Unique_letters)
l = []
m = []
for i in Unique_letters:
    l.append("Jim quickly realized that the beautiful gowns are expensive".count(i))
    
count_letters = dict() 

i = 0
while i < 27:
    count_letters[str(Unique_letters[i])] = l[i]
    i = i + 1
    
del count_letters[" "]        

#### Exercise 1c
- Rewrite your code from 1b to make a function called `counter` that takes a string `input_string` and returns a dictionary of letter counts `count_letters`.
- Use your function to call `counter(sentence)`.

In [6]:
def counter(input_string):
     
    Unique_letters = set(input_string)
    Unique_letters = list(Unique_letters)
    l = []
    for i in Unique_letters:
        l.append(input_string.count(i))
    count_letters = dict() 
    x = 0
    while x < len(Unique_letters):
        count_letters[str(Unique_letters[x])] = l[x]
        x = x + 1
    try:
        for p in (" ",",","-",".","\n"):
            del count_letters[p]
    except:
        pass
    print(count_letters)      
 
counter(sentence)

{'t': 4, 'o': 1, 'h': 2, 'x': 1, 'v': 1, 'r': 2, 'm': 1, 'n': 2, 'f': 1, 'z': 1, 'e': 8, 'p': 1, 'l': 3, 'q': 1, 'c': 1, 'y': 1, 'w': 1, 'J': 1, 'i': 5, 'u': 3, 'd': 1, 'g': 1, 's': 2, 'a': 4, 'b': 1, 'k': 1}


#### Exercise 1d
- Abraham Lincoln was a president during the American Civil War. His famous 1863 Gettysburg Address has been stored as `address`. Use the `counter` function from 1c to return a dictionary consisting of the count of each letter in this address and save it as `address_count`.

In [7]:
address = """Four score and seven years ago our fathers brought forth on this continent, a new nation, 
conceived in Liberty, and dedicated to the proposition that all men are created equal. Now we are engaged in a 
great civil war, testing whether that nation, or any nation so conceived and so dedicated, can long endure. 
We are met on a great battle-field of that war. We have come to dedicate a portion of that field, as a final 
resting place for those who here gave their lives that that nation might live. It is altogether fitting and proper 
that we should do this. But, in a larger sense, we can not dedicate -- we can not consecrate -- we can not hallow -- 
this ground. The brave men, living and dead, who struggled here, have consecrated it, far above our poor power to add 
or detract. The world will little note, nor long remember what we say here, but it can never forget what they did here. 
It is for us the living, rather, to be dedicated here to the unfinished work which they who fought here have thus far so 
nobly advanced. It is rather for us to be here dedicated to the great task remaining before us -- that from these honored 
dead we take increased devotion to that cause for which they gave the last full measure of devotion -- that we here 
highly resolve that these dead shall not have died in vain -- that this nation, under God, shall have a new birth of 
freedom -- and that government of the people, by the people, for the people, shall not perish from the earth."""   

address_count = counter(address)


{'t': 124, 'o': 93, 'h': 80, 'G': 1, 'v': 24, 'B': 1, 'W': 2, 'r': 79, 'I': 3, 'm': 13, 'N': 1, 'n': 76, 'f': 26, 'e': 165, 'p': 15, 'l': 41, 'F': 1, 'q': 1, 'c': 31, 'y': 10, 'w': 26, 'T': 2, 'i': 65, 'L': 1, 'u': 21, 'd': 58, 'g': 27, 's': 44, 'a': 102, 'b': 13, 'k': 3}


#### Exercise 1f
- The frequency of each letter in the Gettysburg Address is already stored as `address_count`. Use this dictionary to find the most common letter in the Gettysburg address.

In [8]:
string = "Jim quickly realized that the beautiful gowns are expensive"

In [9]:
count_string = dict()

In [10]:
for letter in string:
    if letter != " ":
        count_string[letter] = string.count(letter)
    else:
        pass

In [11]:
count_string

{'J': 1,
 'i': 5,
 'm': 1,
 'q': 1,
 'u': 3,
 'c': 1,
 'k': 1,
 'l': 3,
 'y': 1,
 'r': 2,
 'e': 8,
 'a': 4,
 'z': 1,
 'd': 1,
 't': 4,
 'h': 2,
 'b': 1,
 'f': 1,
 'g': 1,
 'o': 1,
 'w': 1,
 'n': 2,
 's': 2,
 'x': 1,
 'p': 1,
 'v': 1}

In [12]:
type(count_string)

dict

### Exercise 2

Consider a circle inscribed in a square. The ratio of their areas (the ratio of the area of the circle to the area of the square) is $\frac{\pi}{4}$. In this six-part exercise, we will find a way to approximate this value.

#### Exercise 2a
- Using the `math` library, calculate and print the value of $\frac{\pi}{4}$

In [13]:
import math
print(math.pi/4)

0.7853981633974483


#### Exercise 2b
- Using `random.uniform()`, create a function `rand()` that generates a single float between $-1$ and $1$.
- Call `rand()` once. For us to be able to check your solution, we will use `random.seed()` to fix the seed value of the random number generator.

In [14]:
import random # imports module "random"

random.seed(1) # Fixes the see of the random number generator.

def rand():
   return(random.uniform(-1,1))

rand()

-0.7312715117751976

#### Exercise 2c
- The distance between two points x and y is the square root of the sum of squared differences along each dimension of x and y. Write a function` distance(x, y)` that takes two vectors as its input and outputs the distance between them. Use your function to find the distance between $x=(0,0)$ and $y=(1,1)$.

In [15]:
def distance(x, y):
    import math
    dist = math.sqrt(((x[0] - y[0])**2) + ((x[1] - y[1])**2))
    return(dist)

distance((0,0), (1,1))

1.4142135623730951

#### Exercise 2d

- Write a function `in_circle(x, origin)` that determines whether a point in a two dimensional plane falls within a unit circle surrounding a given origin.
    - Your function should return a boolean `True` if the distance between `x` and `origin` is less than 1 and `False` otherwise.
    - Use `distance(x, y)` as defined in 2c.
- Use your function to determine whether the point (1,1) lies within the unit circle centered at (0,0).

In [16]:
def in_circle(x, origin = [0,0]):
       import math
       dist = math.sqrt(((x[0] - origin[0])**2) + ((x[1] - origin[1])**2))
       if dist < 1:
            return True
       else:
            return False  

In [17]:
in_circle((1,1))

False

In [18]:
random.seed(1)

In [19]:
R = 10000

random_numbers =[]

for i in range(0,10000):
    random_numbers.append((rand(), rand()))
    
inside = []

for i in random_numbers:
    if in_circle(i) == True:
        inside.append(True)
    else:
        inside.append(False)
        

In [20]:
len(inside)

10000

In [21]:
sum(inside) # Boolean True = 1, False = 0

7790

In [22]:
proportion = sum(inside)/len(inside)

In [23]:
proportion

0.779

**_Code given by edX_**    

`random.seed(1)`

`inside = []`      
`R = 10000`    
`for i in range(R):`      
&emsp; &emsp; &emsp; &emsp; `point = [rand(), rand()]`      
&emsp; &emsp; &emsp; &emsp; `inside.append(in_circle(point))`

`print(sum(inside) / R)`    
`0.779`  

In [24]:
import math

In [25]:
math.pi/4

0.7853981633974483

In [26]:
(math.pi/4) - proportion

0.006398163397448253

In [27]:
random.seed(1)

x = []

for i in range(0,1000):
    x.append(random.uniform(0,1))

In [28]:
def moving_window_average(centre, n_neighbours):
    window_width = (2*n_neighbours+1)  # evenly spaced from either sides of centre
    window_start_index = (centre-n_neighbours)
    window_end_index = (centre+n_neighbours+1) # list slicing will stop 1 index before the terminal index
    window_sum = sum(x[window_start_index:window_end_index])  # total of all elements inside the window
    return window_sum/window_width  # mean of the window
    

In [29]:
moving_window_average(9,5)  # 0.45325045824763405

0.45325045824763405

In [30]:
def MVA(n_neighbours):
    seq = []
    for i in range(0,1000):
        if i < (999-n_neighbours):
            seq.append(sum(x[i : (i+(2*n_neighbours)+1)]) / ((2*n_neighbours)+1))
        else:
            pass
    minimum = min(seq)
    maximum = max(seq)
    return([minimum, maximum])                   

In [31]:
MVA(1)

[0.03816158094808993, 0.9510005995001753]

In [32]:
MVA(7)

[0.3041477707891983, 0.7631568204800142]

In [33]:
for i in range(1,10):
    print("for n_neighbours = ", i, ":", MVA(i))

for n_neighbours =  1 : [0.03816158094808993, 0.9510005995001753]
for n_neighbours =  2 : [0.11050047254034456, 0.9121462444497416]
for n_neighbours =  3 : [0.18268711153577014, 0.8964262339570169]
for n_neighbours =  4 : [0.2363289786358497, 0.8593436734733525]
for n_neighbours =  5 : [0.2686241474150845, 0.7728525560925408]
for n_neighbours =  6 : [0.26122500035438756, 0.7683263756645504]
for n_neighbours =  7 : [0.3041477707891983, 0.7631568204800142]
for n_neighbours =  8 : [0.29942121825755064, 0.7460167136483833]
for n_neighbours =  9 : [0.3035884749888967, 0.7442838713465039]


In [34]:
def MVA_range(n_neighbours):
    seq = []
    for i in range(0,1000):
        if i < (999-n_neighbours):
            seq.append(sum(x[i : (i+(2*n_neighbours)+1)]) / ((2*n_neighbours)+1))
        else:
            pass
    minimum = min(seq)
    maximum = max(seq)
    deviation = maximum - minimum
    return(deviation) 

In [35]:
for i in range(1,10):
    print("for n_neighbours = ", i, "; difference between min & max is", MVA_range(i))

for n_neighbours =  1 ; difference between min & max is 0.9128390185520854
for n_neighbours =  2 ; difference between min & max is 0.801645771909397
for n_neighbours =  3 ; difference between min & max is 0.7137391224212468
for n_neighbours =  4 ; difference between min & max is 0.6230146948375028
for n_neighbours =  5 ; difference between min & max is 0.5042284086774562
for n_neighbours =  6 ; difference between min & max is 0.5071013753101629
for n_neighbours =  7 ; difference between min & max is 0.4590090496908159
for n_neighbours =  8 ; difference between min & max is 0.44659549539083265
for n_neighbours =  9 ; difference between min & max is 0.4406953963576072


#### Exercise 2e

- Create a list `inside` of `R=10000` booleans that determines whether or not a point falls within the unit circle centered at `(0,0)`. 
    - Use the `rand` function from 2b to generate `R` randomly located points.
    - Use the function `in_circle` to test whether or not a given pint falls within the unit circle.
- Find the proportion of points that fall within the circle by summing all `True` values in the `inside` list; then divide the answer by `R` to obtain a proportion.
- Print your answer. This proportion is an estimate of the ratio of the two areas!

In [36]:
random.seed(1) 

# write your code here!


#### Exercise 2f

- Find the difference between your estimate from part 2e and `math.pi / 4`. Note: `inside` and `R` are defined as in Exercise 2e.

In [37]:
# write your code here!


### Exercise 3

A list of numbers representing measurements obtained from a system of interest can often be noisy. One way to deal with noise to smoothen the values by replacing each value with the average of the value and the values of its neighbors.

#### Exercise 3a

- Write a function `moving_window_average(x, n_neighbors)` that takes a list `x` and the number of neighbors `n_neighbors` on either side of a given member of the list to consider.
- For each value in `x`, `moving_window_average(x, n_neighbors)` computes the average of the value and the values of its neighbors.
- `moving_window_average` should return a list of averaged values that is the same length as the original list.
- If there are not enough neighbors (for cases near the edge), substitute the original value for a neighbor for each missing neighbor.
- Use your function to find the moving window sum of `x=[0,10,5,3,1,5]` and `n_neighbors=1`.

In [38]:
def moving_window_average(x, n_neighbors=1):
    n = len(x)
    width = n_neighbors*2 + 1
    x = [x[0]]*n_neighbors + x + [x[-1]]*n_neighbors
    # To complete the function,
    # return a list of the mean of values from i to i+width for all values i from 0 to n-1.

x = [0,10,5,3,1,5]
print(moving_window_average(x, 1))

None


#### Exercise 3b
- Compute and store `R=1000` random values from 0-1 as `x`.
- Compute the moving window average for `x` for values of `n_neighbors` ranging from 1 to 9 inclusive.
- Store `x` as well as each of these averages as consecutive lists in a list called `Y`

In [39]:
random.seed(1) # This line fixes the value called by your function,
               # and is used for answer-checking.
    
# write your code here!


#### Exercise 3c

- For each list in `Y`, calculate and store the range (the maximum minus the minimum) in a new list ranges.
- Print your answer. As the window width increases, does the range of each list increase or decrease? Why do you think that is?

In [40]:
# write your code here!
